File: phone/app/src/main/java/net/novucs/slither/GameSnapshot.java
package net.novucs.slither;

import java.util.Deque;
import java.util.List;

public interface GameSnapshot {

    GameState getState();

    class Connect implements GameSnapshot {

        private final String message;

        public Connect(String message) {
            this.message = message;
        }

        public String getMessage() {
            return message;
        }

        @Override
        public GameState getState() {
            return GameState.CONNECT;
        }
    }

    class Play implements GameSnapshot {

        private final List<Vector2i> rewards;
        private final Deque<Vector2i> player1;
        private final Deque<Vector2i> player2;

        public Play(List<Vector2i> rewards, Deque<Vector2i> player1, Deque<Vector2i> player2) {
            this.rewards = rewards;
            this.player1 = player1;
            this.player2 = player2;
        }

        public List<Vector2i> getRewards() {
            return rewards;
        }

        public Deque<Vector2i> getPlayer1() {
            return player1;
        }

        public Deque<Vector2i> getPlayer2() {
            return player2;
        }

        @Override
        public GameState getState() {
            return GameState.PLAY;
        }
    }

    class Complete implements GameSnapshot {

        private final String message;

        public Complete(String message) {
            this.message = message;
        }

        public String getMessage() {
            return message;
        }

        @Override
        public GameState getState() {
            return GameState.COMPLETE;
        }
    }
}

File: android/app/src/main/java/net/novucs/slither/GameSnapshot.kt
package net.novucs.slither

import java.util.*

/**
 * An immutable snapshot of the game, used for rendering.
 */
interface GameSnapshot {

    val state: GameState

    /**
     * Displays the reconnecting message when in the connection state.
     */
    class Connect(val message: String) : GameSnapshot {
        override val state: GameState
            get() = GameState.CONNECT
    }

    /**
     * Displays all rewards, players and their scores when in play state.
     */
    class Play(val rewards: List<Vector2i>,
               val player1: Deque<Vector2i>,
               val player2: Deque<Vector2i>,
               val score1: Int,
               val score2: Int) : GameSnapshot {
        override val state: GameState
            get() = GameState.PLAY
    }

    /**
     * Displays a completion message and player scores when in complete state.
     */
    class Complete(val message: String,
                   val score1: Int,
                   val score2: Int) : GameSnapshot {
        override val state: GameState
            get() = GameState.COMPLETE
    }
}
